src/libraries/Microsoft.PowerFx.Core/Public/Engine.cs

```diff
@@ -173,92 +201,115 @@ public CheckResult Check(string expressionText, RecordType parameterType, Parser
+        internal CheckResult Check(ParseResult parse, ParserOptions options)
+        {
+            parse.Options = options;
+            return Check(parse, null, options);
+        }
+
+        public CheckResult Check(ParseResult parse, RecordType parameterType = null)
         {
-            var bindingConfig = new BindingConfig(allowsSideEffects: options?.AllowsSideEffects == true);
-            var symbols = SymbolTable.NewFromRecord(parameterType);
+            var check = new CheckResult(this)
+               .SetText(parse)
+               .SetBindingInfo(parameterType);
 
-            return CheckInternal(parse, bindingConfig, symbols);
+            CheckWorker(check);
+            return check;
         }
 
         public CheckResult Check(
             string expressionText,
             ParserOptions options = null,
             ReadOnlySymbolTable symbolTable = null)
         {
-            var parse = Parse(expressionText, options);
-            return Check(parse, options, symbolTable);
+            var check = new CheckResult(this)
+                .SetText(expressionText, options)
+                .SetBindingInfo(symbolTable);
+
+            CheckWorker(check);
+            return check;
         }
 
-        public CheckResult Check(
-            ParseResult parse,
-            ParserOptions options = null,
-            ReadOnlySymbolTable symbolTable = null)
+        // Apply a standard set of operations on the CheckResult.
+        // If callers want more granularity, they can create the CheckResult themselves. 
+        private void CheckWorker(CheckResult check)
         {
-            var bindingConfig = new BindingConfig(allowsSideEffects: options?.AllowsSideEffects == true);
+            check.ApplyBindingInternal();
+            check.ApplyErrors();
+            check.ApplyDependencyAnalysis();
+        }
 
-            return CheckInternal(parse, bindingConfig, symbolTable);
+        // Called after check result, can inject additional errors or constraints. 
+        protected virtual IEnumerable<ExpressionError> PostCheck(CheckResult check)
+        {
+            return new ExpressionError[0];
         }
 
-        private CheckResult CheckInternal(
-            ParseResult parse,
-            BindingConfig bindingConfig = null,
-            ReadOnlySymbolTable symbolTable = null)
+        internal IEnumerable<ExpressionError> InvokePostCheck(CheckResult check)
+        {
+            return this.PostCheck(check);
+        }
+
+        // Setting rule sope which will get passed into Binder. 
+        // Prefer to avoid this hook and use SymbolTables instead. 
+        private protected virtual RecordType GetRuleScope()
+        {
+            return null;
+        }
+
+        private BindingConfig GetDefaultBindingConfig()
+        {
+            var ruleScope = this.GetRuleScope();
+            bool useThisRecordForRuleScope = ruleScope != null;
```

Comment:
>useThisRecordForRuleScope [](http://example.com/codeflow?start=17&length=25)

I don't see a reason for creating a variable which is only used once.
Logically, I'm wondering if BindingConfig.Default going to modify ruleScope?

Reply:
It's a style thing - it makes the code more self-documenting.